qiskit/circuit/quantumcircuit.py

```diff
@@ -1396,6 +1396,26 @@ def decompose(
         decomposed_dag = pass_.run(circuit_to_dag(self))
         return dag_to_circuit(decomposed_dag)
 
+    def translate(self, basis):
+        """Call a basis translation pass on this circuit to translate to
+        a desired gate set.
+
+        Args:
+            basis (list(str)): List of target gates.
+
+        Returns:
+            QuantumCircuit: translated circuit
+        """
+        # pylint: disable=cyclic-import
+        from qiskit.transpiler.passes.basis.basis_translator import BasisTranslator
+        from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
+        from qiskit.converters.circuit_to_dag import circuit_to_dag
+        from qiskit.converters.dag_to_circuit import dag_to_circuit
+
+        pass_ = BasisTranslator(sel, basis)
+        translated_dag = pass_.run(circuit_to_dag(self.decompose()))
```

Comment:
This is better, but I think will still miss cases where (e.g. for `Isometry`) we have multiple levels of unrolling that would need to be done. Come to think of it, is there a reason not to do something like:
```
from qiskit.compiler import transpile
return transpile(self, basis_gates=basis, optimization_level=0)
```
? This should cover all the edge cases and be easier to keep up to date in the future.

Reply:
This seems good although does it reduce the need for a convenience function since it's already pretty simple? Then we have a situation where a higher level function is called by what might be considered a lower level one.

Reply:
I don't think so, because this seems like a common enough use-case and the discoverability of this method is substantially better than the equivalent `transpile` call. I could see this leading to some unusual calls (like if a transpiler pass were to internally call `.translate` instead of delegating to the appropriate pass) but I think `QuantumCircuit` is already the high level interface (and `DAGCircuit` the lower level) that I can't think of any disadvantage in that.

I'm not sure about `decompose`, because it has some specific behavior (only wanting to unroll a single level) that can't be achieved with a single `transpile` call. However, the `.translate` method I think covers most of the use cases for `.decompose` such that we can look at potentially deprecating it in the future.